cmd/exo/apply.go

```diff
@@ -58,23 +59,52 @@ var applyCmd = &cobra.Command{
 }
 
 func apply(ctx context.Context, kernel api.Kernel, workspace api.Workspace, args []string) error {
-	input := &api.ApplyInput{}
+	var manifestPath string
+	workingDir, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("getting working dir: %w", err)
+	}
+
 	if len(args) > 0 {
-		manifestPath := args[0]
-		input.ManifestPath = &manifestPath
-
-		// We're not necessarily in the workspace root here,
-		// so send the file contents too.
-		bs, err := ioutil.ReadFile(manifestPath)
-		if err != nil {
-			return fmt.Errorf("reading manifest file: %w", err)
+		manifestPath = args[0]
+		if !path.IsAbs(manifestPath) {
+			manifestPath = path.Join(workingDir, manifestPath)
 		}
-		s := string(bs)
-		input.Manifest = &s
 	}
-	if applyFlags.Format != "" {
-		input.Format = &applyFlags.Format
+
+	for _, candidate := range manifestCandidates {
```

Comment:
Why is this search part of the CLI? Seems like as much of this logic as possible should be part of the server, so that the GUI can reuse the logic.

Reply:
Yes I think there's follow up work to create a `get-manifest-candidates` command on the server that returns a list of possible manifests in priority order.

Reply:
It's not obvious from looking at the changes: Why does the CLI need this at all?

Reply:
Given the assumption that everyone will have an exo file (which is perhaps an unfinished discussion), I think it makes sense that the `apply` endpoint takes just one format of exo file and that its responsibility is to simply apply it, not to possibly convert it, save it  and then apply a converted manifest. In order to achieve that, I moved the conversion and saving steps into the CLI. I'm also fine with having a `convert` endpoint that does the conversion.

The actual locating of the file feels like a client concern: for the CLI running a manifest in the working directory seems like a reasonable choice but for the UI you might want to have a `run` button which gives you a choice between manifests. Something like this:

![](https://i.stack.imgur.com/gAt78.png)

Reply:
> Given the assumption that everyone will have an exo file

let's discuss more via slack

> The actual locating of the file feels like a client concern

I think that some change in perspective is necessary here. The word "client" is sort of ambiguous/meaningless in the context of exo. The CLI tool is a user interface. In terms of separation of concerns, it should only be concerned with command parsing, terminal rendering, etc. From this perspective, searching for manifest files is a _workspace_ concern, and therefore should be centralized where any UI (ie the CLI or GUI) can access it.

> a run button which gives you a choice between manifests

I think about it as having a choice of run configurations - probably called stacks - instead of a choice of manifests. Seems like stack = manifest + config.

Reply:
back to the manifest search logic:

One bit of feedback we've received is that somebody didn't realize they had to run `apply` after their manifest changed. We're also planning on adding a file watcher anyway, so it would be good that if a file changes, the server knows if it is or could potentially be a manifest file and some UI shows that there is a diff between the state and the manifest. Then you'd want to either commit the change to the manifest, or apply the manifest to update the state.

Reply:
> searching for manifest files is a workspace concern

Yes, but figuring out what to do with them once you've found them feels like a UI concern. In the CLI you probably want to prefer the one in your working directory. In the GUI you probably want to choose the one in the workspace root. So I'd propose a `get-manifest-candidates` which returns all possible manifests and then the GUI or the CLI makes the final choice about what to do.

Reply:
Fair point, I had not considered working directory vs workspace root. Though it's not yet clear to me how that should work, especially if we get in to hierarchical components for subsystems and the like.

I'm not opposed to the `get-manifest-candidates` API in general, my concern is the remaining extra logic now of the conversion, etc happening in the CLI. Can we do some design work on the UI so we know how this is going to work there? I want to make sure we have really strong CLI/GUI parity & I suspect that this will need to be refactored back closer to the original separation of concerns pretty much immediately upon trying to add an "apply" feature to the UI.